File: neet_code_all_cpp/array_hash/design_parking_system.cpp
#include <iostream>
#include <unordered_map>
#include <array>

// AC
class ParkingSystem {
    private:
        int max_big;
        int max_medium;
        int max_small;
        int n_big;
        int n_medium;
        int n_small;

    public:
        ParkingSystem(int big, int medium, int small)
            :max_big(big), max_medium(medium), max_small(small),
            n_big(0), n_medium(0), n_small(0)
        {}

        bool addCar(int carType) {
            switch(carType) {
                case 1:
                    {
                        if(n_big >= max_big) {
                            return false; 
                        } else {
                            ++n_big;
                            break;
                        }
                    }
                case 2:
                    {
                        if(n_medium >= max_medium) {
                            return false; 
                        } else {
                            ++n_medium;
                            break;
                        }
                    }

                default:
                    {
                        if(n_small >= max_small) {
                            return false; 
                        } else {
                            ++n_small;
                            break;
                        }
                    }
            }

            return true;
        }
};

// AC
class ParkingSystemAnsPython {
    private:
        std::unordered_map<int, std::array<int, 2>> parking;

    public:
        ParkingSystemAnsPython(int big, int medium, int small) {
            parking = {
                {1, {0, big}},
                {2, {0, medium}},
                {3, {0, small}}
            };
        }

        bool addCar(int carType) {
            if(parking[carType][0] + 1 <= parking[carType][1]) {
                ++parking[carType][0];
                return true;
            }

            return false;
        }
};

int main(void) {
    ParkingSystem p_0(1, 1, 0);
    std::cout << p_0.addCar(1) << std::endl; 
    std::cout << p_0.addCar(2) << std::endl; 
    std::cout << p_0.addCar(3) << std::endl; 
    std::cout << p_0.addCar(1) << std::endl; 

    ParkingSystemAnsPython p_py(1, 1, 0);
    std::cout << p_py.addCar(1) << std::endl; 
    std::cout << p_py.addCar(2) << std::endl; 
    std::cout << p_py.addCar(3) << std::endl; 
    std::cout << p_py.addCar(1) << std::endl; 
}

File: neet_code_all/src/array_hash/design_parking_system.rs
use std::collections::HashMap;

// AC
struct ParkingSystem {
    max_big: i32,
    n_big: i32,
    max_medium: i32,
    n_medium: i32,
    max_small: i32,
    n_small: i32,
}

impl ParkingSystem {
    pub fn new(big: i32, medium: i32, small: i32) -> Self {
        Self {
            max_big: big,
            max_medium: medium,
            max_small: small,
            n_big: 0,
            n_medium: 0,
            n_small: 0,
        }
    }

    pub fn add_car(&mut self, car_type: i32) -> bool {
        match car_type {
            1 => {
                if self.n_big >= self.max_big {
                    return false;
                }
                self.n_big += 1;
            }
            2 => {
                if self.n_medium >= self.max_medium {
                    return false;
                }
                self.n_medium += 1;
            }
            _ => {
                if self.n_small >= self.max_small {
                    return false;
                }
                self.n_small += 1;
            }
        }

        true
    }
}

// Pyrhonの模範解答より
// AC
struct SolutionAnsPython {
    parking: HashMap<i32, [i32; 2]>,
}

impl SolutionAnsPython {
    pub fn new(big: i32, medium: i32, small: i32) -> Self {
        Self {
            parking: HashMap::from([(1, [0, big]), (2, [0, medium]), (3, [0, small])]),
        }
    }

    pub fn add_car(&mut self, car_type: i32) -> bool {
        if self.parking[&car_type][0] + 1 <= self.parking[&car_type][1] {
            if let Some(num_car) = self.parking.get_mut(&car_type) {
                num_car[0] += 1;
            }
            return true;
        }

        false
    }
}

fn main() {
    let mut p_0 = ParkingSystem::new(1, 1, 0);
    println!("p_0.add_car(1): {}", p_0.add_car(1));
    // => true
    println!("p_0.add_car(2): {}", p_0.add_car(2));
    // => true
    println!("p_0.add_car(3): {}", p_0.add_car(3));
    // => false
    println!("p_0.add_car(1): {}", p_0.add_car(1));
    // => false

    let mut p_py = SolutionAnsPython::new(1, 1, 0);
    println!("p_py.add_car(1): {}", p_py.add_car(1));
    // => true
    println!("p_py.add_car(2): {}", p_py.add_car(2));
    // => true
    println!("p_py.add_car(3): {}", p_py.add_car(3));
    // => false
    println!("p_py.add_car(1): {}", p_py.add_car(1));
    // => false
}
